include/mbedtls/pk.h

```diff
@@ -116,42 +131,108 @@ typedef struct
 #define MBEDTLS_PK_DEBUG_MAX_ITEMS 3
 
 /**
- * \brief           Public key information and operations
+ * \brief           Key pair information and operations
  */
 typedef struct mbedtls_pk_info_t mbedtls_pk_info_t;
 
 /**
- * \brief           Public key container
+ * \brief           Key pair container
  */
 typedef struct
 {
-    const mbedtls_pk_info_t *   pk_info; /**< Public key informations        */
-    void *                      pk_ctx;  /**< Underlying public key context  */
+    const mbedtls_pk_info_t *   pk_info; /**< Algorithm information          */
+    void *                      pk_ctx;  /**< Underlying key pair context */
 } mbedtls_pk_context;
 
+/**
+ * \brief           Access the type name
+ *
+ * \param ctx       Context to use
+ *
+ * \return          Type name on success, or "invalid PK"
+ */
+const char * mbedtls_pk_get_name( const mbedtls_pk_context *ctx );
+
+/**
+ * \brief           Get the key type
+ *
+ * \param ctx       Context to use
+ *
+ * \return          Type on success, or MBEDTLS_PK_NONE
+ *
+ * \note            This function returns the type of the key pair object. The
+ *                  type encodes the representation of the object as well as
+ *                  the operations that it can be used for. To test whether
+ *                  the object represents a key of a recognized type such
+ *                  as RSA or ECDSA, call \c mbedtls_pk_can_do().
+ */
+mbedtls_pk_type_t mbedtls_pk_get_type( const mbedtls_pk_context *ctx );
+
+/**
+ * \brief           Merge key types with the same representation
+ *
+ * \param type      Any key type
+ * \return          A canonical representative among the types with the
+ *                  same key representation. This is \c MBEDTLS_PK_RSA
+ *                  for RSA keys using the built-in software engine and
+ *                  MBEDTLS_PK_ECKEY for EC keys using the built-in
+ *                  software engine. Note that for keys of type
+ *                  \c MBEDTLS_PK_OPAQUE, the type does not specify the
+ *                  representation.
+ */
+static inline mbedtls_pk_type_t mbedtls_pk_representation_type( mbedtls_pk_type_t type )
```

Comment:
We should make sure that this function is used everywhere where previously the representation type was inferred directly, for example [here](https://github.com/ARMmbed/mbedtls/blob/development/library/x509write_crt.c#L325).

Reply:
Your link points to a use of `can_do`, and this function already does the right thing. Did you mean a different location? There shouldn't be much use for `mbedtls_pk_representation_type` or `mbedtls_pk_type`: in most cases the right thing is to call `mbedtls_pk_can_do`.

Reply:
Yes, that was not accurate, I thought of [this](https://github.com/ARMmbed/mbedtls/blob/development/library/pkwrite.c#L155) instead, where public key contexts get exported to DER. In this situation, success or failure should be determined by the representation type.

Reply:
Indeed. I wonder if we need to deprecate `mbedtls_pk_get_type`. When would users use it? Tracking this internally in IOTSSL-2054

Reply:
@gilles-peskine-arm https://github.com/ARMmbed/mbedtls/blob/development/library/ssl_srv.c#L658 is another example where a representation-type check is a useful safeguard.

Reply:
@gilles-peskine-arm @hanno-arm I think another example where using `mbedtls_pk_representation_type()` would probably be better than the existing code is `x509_profile_check_key()` in `x509_crt.c`.

Reply:
@mpg Could you add that to IOTSSL-2054?

Reply:
Done.